### PR TITLE
add flag for building bluechi-selinux with any port

### LINKED ISSUE #836: Remove the requirement on policycoreutils-python-utils for the bluechi-selinux package
This is a continuation of https://github.com/containers/qm/issues/350

We need to get rid of the `policycoreutils-python-utils` in the `bluechi-selinux` package

> Can we just eliminate the lines about ports. semanage port This is pulling in the policycoreutils-python package.
We should just loosen the policy and allow bluechi to use all ports. Perhaps add a boolean and just turn this on for RHIVOS.

### PR DESCRIPTION
**Pull request by engelmi**
Fixes: https://github.com/eclipse-bluechi/bluechi/issues/836

Previously, the SELinux policy for BlueChi only contained one type for the port and was used only by the bluechi-agent. This adds a new port type and and uses one for the bind of bluechi-controller and another for the connect of bluechi-agent.
Based on this, tunable policies are added to turn on/off the option to use any port for both components. 
In addition, it removes the implicit permissive policy so that violations to the policies are enforced by default (can be set permissive later on by the user again).

This PR also updates the RPM spec of BlueChi so that it does not require the semange tool and thus is not pulling in python as a dependency anymore. It adds also another option to the build to enable any port policy for bluechi-controller an bluechi-agent.

It also adds documentation about BueChi's SELinux policy, restructuring the order of the table of contents and introducing the section `Security` which contains the selinux policy and mTLS double proxy approach.

__NOTE:__

The changes for the documentation look like this:
image

### CODE REVIEW

**Review by mwperina — APPROVED** [review]
LGTM

**Comment by engelmi** [comment]
@rhatdan PTAL

File: doc/docs/security/selinux.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
+<!-- markdownlint-disable-file MD013-->
+# BlueChi's SELinux policy
+
+BlueChi provides a custom SELinux policy, limiting access of the `bluechi-controller` and `bluechi-agent`. It can be installed via
+
+```bash
+dnf install bluechi-selinux
+```
+
+By default, the package adds an allow rule for the `bluechi-agent` to connect to BlueChi's default port 842. If the controller port where the agent is connecting to changes, then an AVC entry is created:
```
Comment (mwperina) [inline doc/docs/security/selinux.md]: I'd add some commas for better clarity:

"If the controller port, where the agent is connecting to, changes then an AVC entry is created:"

File: doc/docs/security/selinux.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+```bash
+time->Thu Mar 28 14:12:35 2024
+type=PROCTITLE msg=audit(1711635155.532:22): proctitle="/usr/libexec/bluechi-agent"
+type=SYSCALL msg=audit(1711635155.532:22): arch=c000003e syscall=42 success=no exit=-115 a0=a a1=555f97c27c50 a2=10 a3=7ffd4b114d74 items=0 ppid=1 pid=531 auid=4294967295 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none) ses=4294967295 comm="bluechi-agent" exe="/usr/libexec/bluechi-agent" subj=system_u:system_r:bluechi_agent_t:s0 key=(null)
+type=AVC msg=audit(1711635155.532:22): avc:  denied  { name_connect } for  pid=531 comm="bluechi-agent" dest=8450 scontext=system_u:system_r:bluechi_agent_t:s0 tcontext=system_u:object_r:unreserved_port_t:s0 tclass=tcp_socket permissive=1
+```
+
+This rule, however, is not enforcing, but permissive and thus not preventing the `bluechi-agent` to connect to the different host.
```
Comment (mwperina) [inline doc/docs/security/selinux.md]: I'd rephrase:

"However this rule is not enforcing, but permissive, and thus not preventing the `bluechi-agent` to connect to the different host."

Comment (engelmi) [reply doc/docs/security/selinux.md]: Changed.

File: doc/docs/security/selinux.md (hunk 1 above)
Comment (engelmi) [reply doc/docs/security/selinux.md]: I figured out a way to link to the man page section for the `ControllerPort`. I rephrased and simplified the sentence it a bit.

Comment (mwperina) [reply doc/docs/security/selinux.md]: OK

**Review by mwperina — APPROVED** [review]

**Comment by alexlarsson** [comment]
This doesn't seem right to me. The selinux module still references the bluechi_port_t type, and if the semanage call was never called, then this will not be set to anything. I don't know selinux that well, so I'm not sure if that will lead to some error, or if the port is assumed to be zero or something.

Either way, unless I'm totally off, there is no way this can correctly work.

I think we need something like systemd_socket_proxyd_bind_any in 
https://github.com/fedora-selinux/selinux-policy/blob/85bbb6d1bc69e84ea2440cff647e426b74f1b40a/policy/modules/system/systemd.te

This has both a port type which can be assigned with semanage (systemd_socket_proxyd_port_t) and boolean options ( systemd_socket_proxyd_bind_any and systemd_socket_proxyd_connect_any) which instead makes the rules work on any port.  I think the right approach is to use something like that with the boolean options set to true by default, and then have the optional rpm post-scripts set both the port and the boolean.

**Comment by alexlarsson** [comment]
> This doesn't seem right to me. The selinux module still references the bluechi_port_t type, and if the semanage call was never called, then this will not be set to anything. I don't know selinux that well, so I'm not sure if that will lead to some error, or if the port is assumed to be zero or something.

I guess what happen is that no port will ever get that label assigned, so no port will ever pass the rules.

**Comment by engelmi** [comment]
> > This doesn't seem right to me. The selinux module still references the bluechi_port_t type, and if the semanage call was never called, then this will not be set to anything. I don't know selinux that well, so I'm not sure if that will lead to some error, or if the port is assumed to be zero or something.
> 
> I guess what happen is that no port will ever get that label assigned, so no port will ever pass the rules.

The port would need to be assigned to that label with `semanage` later on. Till then, basically no port is assigned and an AVC is always created - but the name bind itself is not blocked. 

I'll have a look at the `systemd_socket_proxyd_bind_any`.

**Comment by alexlarsson** [comment]
> The port would need to be assigned to that label with semanage later on. Till then, basically no port is assigned and an AVC is always created - but the name bind itself is not blocked.

I feel this being permissive like this makes the whole selinux module not very useful though.

**Comment by engelmi** [comment]
> I feel this being permissive like this makes the whole selinux module not very useful though.

I agree. In addition, I think it can't be changed from permissive to enforcing later on (since it wasn't added by semanage according to [the man page]). 

I did a bunch of changes, incl. the tunable policy for the ports based on the systemd example. 
Could you have another look? @alexlarsson

File: doc/docs/security/selinux.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+dnf install bluechi-selinux
+```
+
+By default, the package allows `bluechi-controller` to bind and `bluechi-agent` to connect to any port so that changes of the [ControllerPort](../man/bluechi-agent-conf.md#controllerport-uint16_t) are not being blocked.
+
+In order to allow BlueChi to only use one port, the usage of any needs to be disabled with [setsebool](https://linux.die.net/man/8/setsebool):
+
+```bash
+# Turn off the any policy for bluechi-controller
+setsebool -P bluechi_port_bind_any 0
```
Comment (mwperina) [inline doc/docs/security/selinux.md]: Wouldn't  `bluechi_controller_bind_any_port` be a more descriptive name?

File: doc/docs/security/selinux.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+# Turn off the any policy for bluechi-agent
+setsebool -P bluechi_agent_port_connect_any 0
```
Comment (mwperina) [inline doc/docs/security/selinux.md:21]: Wouldn't be `bluechi_agent_connect_to_any_port` be a more descriptive name?

File: selinux/bluechi.te [hunk 5]
```diff
@@ -16,8 +16,11 @@ init_daemon_domain(bluechi_agent_t, bluechi_agent_exec_t)
 type bluechi_port_t alias hirte_port_t;
 corenet_port(bluechi_port_t)
 
-permissive bluechi_t;
-permissive bluechi_agent_t;
+type bluechi_agent_port_t;
+corenet_port(bluechi_agent_port_t)
+
+gen_tunable(bluechi_port_bind_any, false)
+gen_tunable(bluechi_agent_port_connect_any, false)
```
Comment (mwperina) [inline selinux/bluechi.te]: Could you please move those boolean options to the top of the file and add a description for them similar as system has?

https://github.com/fedora-selinux/selinux-policy/blob/85bbb6d1bc69e84ea2440cff647e426b74f1b40a/policy/modules/system/systemd.te#L14

File: bluechi.spec.in [hunk 6]
```diff
@@ -167,27 +173,21 @@ SELinux policy associated with the bluechi and bluechi-agent daemons
 %post selinux
 # Remove hirte policy
 if [ $1 -eq 1 ]; then
-   semanage port -N -d -p udp 842 2>/dev/null || true
-   semanage port -N -d -p tcp 842 2>/dev/null || true
```
Comment (mwperina) [inline bluechi.spec.in:171]: Don't we need to keep selinux upgrade from hirte to bluechi till Fedora 39 deprecation? I thought that we would want to add a build option to be able to turn it off on selected platforms?

Also, have you tried upgrade from bluechi 0.7 to bluechi 0.8 with any port policy update? Don't we need to remove those specific port options before installing new policy with any port?

File: doc/docs/security/selinux.md (hunk 3 above)
Comment (engelmi) [reply doc/docs/security/selinux.md]: I added the `_controller_` substring, but would keep the `_bind_any` part similar to the [systemd_socket_proxyd_bind_any](https://github.com/fedora-selinux/selinux-policy/blob/85bbb6d1bc69e84ea2440cff647e426b74f1b40a/policy/modules/system/systemd.te#L14)

File: doc/docs/security/selinux.md (hunk 4 above)
Comment (engelmi) [reply doc/docs/security/selinux.md:21]: I would keep the `_connect_any` part similar to the [systemd_socket_proxyd_connect_any](https://github.com/fedora-selinux/selinux-policy/blob/85bbb6d1bc69e84ea2440cff647e426b74f1b40a/policy/modules/system/systemd.te#L22C13-L22C46)

File: selinux/bluechi.te (hunk 5 above)
Comment (engelmi) [reply selinux/bluechi.te]: Done.

File: bluechi.spec.in (hunk 6 above)
Comment (engelmi) [reply bluechi.spec.in:171]: These two lines for semanage basically remove the allow rule for tcp and udp on port 842 - no specific domain is given. I don't think these are needed even for the old hirte considering the latest changes. 

> Also, have you tried upgrade from bluechi 0.7 to bluechi 0.8 with any port policy update? Don't we need to remove those specific port options before installing new policy with any port?

Based on how the any port is implemented:
```
tunable_policy(`bluechi_agent_port_connect_any',`
	corenet_tcp_connect_all_ports(bluechi_agent_t)
',`
	allow bluechi_agent_t bluechi_agent_port_t:tcp_socket name_connect;
')
```
Either the labelled port or the any connect/bind is used. Therefore, even though the port labels aren't removed, those don't have any effect till the any is disabled. And even then, the ports would be "pre-configured" - which is ok, I think, considering the new documentation would guide the user to add ports (and thus viewing these). 

I tried verifying this by upgrading from v0.7 to v0.8 with the any port policy and it really kept the port rules:
```bash
[root@localhost tmp]# semanage port -l | grep bluechi
bluechi_port_t                 tcp      842
bluechi_port_t                 udp      842
```
And starting BlueChi on another port was still possible with the any policy active. Switch it off and running the controller on another port, e.g. 8420, was blocked as expected.

File: doc/docs/security/selinux.md (hunk 3 above)
Comment (mwperina) [reply doc/docs/security/selinux.md]: OK

File: doc/docs/security/selinux.md (hunk 4 above)
Comment (mwperina) [reply doc/docs/security/selinux.md:21]: OK

File: bluechi.spec.in (hunk 6 above)
Comment (mwperina) [reply bluechi.spec.in:171]: OK, makes sense

**Review by mwperina — APPROVED** [review]

**Review by mwperina — CHANGES_REQUESTED** [review]
Just noticed that we forgot to remove dependency on policycoreutils-python-utils

**Review by mwperina — APPROVED** [review]

File: bluechi.spec.in [hunk 7]
```diff
@@ -167,27 +171,21 @@ SELinux policy associated with the bluechi and bluechi-agent daemons
...
    semodule -N -X 200 -r hirte 2>/dev/null || true
 fi
 %selinux_modules_install -s %{selinuxtype} %{_datadir}/selinux/packages/bluechi.pp.bz2
 restorecon -R %{_bindir}/bluechi* &> /dev/null || :
-semanage port -a -t bluechi_port_t -p udp 842 2>/dev/null || semanage port -m -t bluechi_port_t -p udp 842
-semanage port -a -t bluechi_port_t -p tcp 842 2>/dev/null || semanage port -m -t bluechi_port_t -p tcp 842
+%if %{with_selinux_port_any}
+   setsebool -P bluechi_controller_port_bind_any 1
```
Comment (rhatdan) [inline bluechi.spec.in]: combine these lines together, to save install time.

  setsebool -P bluechi_controller_port_bind_any=1 bluechi_agent_port_connect_any=1

File: selinux/bluechi.te [hunk 8]
```diff
@@ -5,6 +5,12 @@ policy_module(bluechi, 1.0.0)
 # Declarations
 #
 
+# Allow bluechi_t to bind to any port instead of one labelled with bluechi_port_t.
+gen_tunable(bluechi_controller_port_bind_any, false)
```
Comment (rhatdan) [inline selinux/bluechi.te]: Why not change the default on these to true and then allow users to turn up the security rather then forcing everyone to turn down the security.

Comment (alexlarsson) [reply selinux/bluechi.te]: Yeah, i think it makes sense to do this, otherwise the default isn't gonna work. If this defaults to true, then only in the case where you want to limit to a port do you have to change anything (setting both any to false and the port).

**Comment by alexlarsson** [comment]
Other than the default change pointed out by dan i think this looks good.

File: selinux/bluechi.te (hunk 8 above)
Comment (engelmi) [reply selinux/bluechi.te]: Good point! I changed both anys to default to `true`.

File: bluechi.spec.in (hunk 7 above)
Comment (engelmi) [reply bluechi.spec.in]: Good point, changed it.

File: README.developer.md [hunk 9]
```diff
@@ -377,9 +377,12 @@ The package build process can be adjusted using following environment variables:
 - `SKIP_BUILDDEP`
   - To skip installation of build dependencies this option should contain `yes` value.
 - `WITH_PYTHON`
-  - To skip building python bluechi modules this option should contain `0`.
+  - To skip building python bluechi modules this option should contain `0`. Default: `1`.
 - `WITH_COVERAGE`
-  - To start collecting coverage this option should contain `1`.
+  - To start collecting coverage this option should contain `1`. Default: `0`.
+- `WITH_SELINUX_PORT_ANY`
```
Comment (mwperina) [inline README.developer.md]: Do we need that build option when both `bluechi_controller_port_bind_any` and `bluechi_agent_port_connect_any` are enabled by default? Do we really need to rebuild the package when we want to limit access to specific port only? Can't limitation to specific port be achieved only after installation by turning off those selinux booleans and run semanage?

Comment (engelmi) [reply README.developer.md]: Good point! I removed it and verified that the default in the `bluechi.te` works as expected. Removed the `with_selinux_port_any` option from the spec file.

File: bluechi.spec.in [hunk 10]
```diff
@@ -167,27 +166,17 @@ SELinux policy associated with the bluechi and bluechi-agent daemons
 %post selinux
 # Remove hirte policy
```
Comment (mwperina) [inline bluechi.spec.in:167]: Please remove also above comment

File: doc/docs/security/selinux.md [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+```bash
+# Set the allowed port for bluechi-controller
+semanage port -a -t bluechi_port_t -p tcp <port>
+
+# Set the allowed port for bluechi-agent
+semanage port -a -t bluechi_agent_port_t -p tcp <port>
+```
+
+## Change from enforcing to permissive
```
Comment (mwperina) [inline doc/docs/security/selinux.md:34]: I think we are trying to avoid permissive mode in general, so is it worth to have that section?

Comment (engelmi) [reply doc/docs/security/selinux.md:34]: I think its not bad to have it. Its not much effort and if somebody for some reason really wants to switch it to permissive, it briefly described here.

**Comment by engelmi** [comment]
Thank you for your help! @alexlarsson @rhatdan @mwperina 
Merging.
